core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveTrivialFilters.java

```diff
@@ -46,10 +68,35 @@ public Result apply(FilterNode filterNode, Captures captures, Context context)
             return Result.ofPlanNode(filterNode.getSource());
         }
 
-        if (predicate.equals(FALSE_LITERAL) || predicate instanceof NullLiteral) {
+        if (predicate.equals(FALSE_LITERAL) || isEffectivelyNull(predicate, context)) {
             return Result.ofPlanNode(new ValuesNode(context.getIdAllocator().getNextId(), filterNode.getOutputSymbols(), emptyList()));
         }
 
         return Result.empty();
     }
+
+    private boolean isEffectivelyNull(Expression expression, Context context)
+    {
+        if (expression instanceof NullLiteral) {
+            return true;
+        }
+
+        Session session = context.getSession();
+        if (isEffectivelyLiteral(plannerContext, session, expression)) {
+            TypeProvider typeProvider = context.getSymbolAllocator().getTypes();
+            Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, typeProvider, expression);
+            Type type = expressionTypes.get(NodeRef.of(expression));
+
+            Object value = evaluateConstantExpression(
+                    expression,
+                    type,
+                    plannerContext,
+                    session,
+                    new AllowAllAccessControl(),
+                    ImmutableMap.of());
```

Comment:
It shouldn't be necessary to evaluate the expression here. That should be taken care of by other optimizers such as SimplifyExpressions.

Instead, just match a NullLiteral or a Cast with a NullLiteral. Those are the only two forms for a simplified null literal.

Reply:
Done